### PR TITLE
ArCargo: Adds a stock market system for buying and selling minerals.

### PR DESCRIPTION
**Pull request by ArcaneMusic**
Alright, technically, it's not a stock market. It's a commodity market, actually, but this is catchier and gets the idea across faster.

## About The Pull Request
image
This PR adds a new machine that can be bought called the **Galactic Mineral Market** (GMM). The Galactic Mineral Market (GMM) allows you to buy and sell minerals wholesale from the market machine. It goes something like this:

### 1. Getting the Machine:
The GMM can be bought as an un-assembled machine for 600 credits from cargo. It's low, but its not mapped in standard as if you're confident in your miners, you shouldn't necessarily need one for the department. Otherwise, it's available for other crewmembers to buy and sell from at their expense.

Cargo pack comes will all supplies necessary to finish the machine. Tools not included.

### 2. Buying Low
Using the machine's UI, you can see all traded minerals and their associated prices. Buy prices are played straight, and can be bought for material price * quantity. The order is then instantly placed on the cargo shuttle, and will be deducted from the buyer's account on shuttle send.

Cargo staff with standard cargo access may toggle the machine to order directly from the cargo budget. Otherwise, materials are only purchased from private supplies.

All purchases are treated as private and must be opened by the recipient like private orders.

### 3. Selling High

It's stocks time. To sell minerals on the market, simply insert any relevant metal stock into the machine. This produces a totally original and not-a-bounty-cube stock block, which can be sold on the cargo shuttle for cargo funds. Stock blocks can also be price-tagged as well following their standard process.

Sold cargo stocks are subject to a 20% processing fee as part of the galactic mineral market.

### Other notes:
The market does not cover all minerals, partially for consistency purposes as well as for balance reasons. Plasma, being a unique material that only Nanotrasen has their hands on, is the sole exporter of plasma in the system, so it stands to reason that it's not on the market, and remains a solo export. Bananium, as it has a rare and expensive conversion rate, works the opposite way, and as such isn't listed on the market either, with the sole source being single cargo sales. All others just don't make much sense to include into the market at this second, so I left them out for now.

**Alright now below I'm going to cover the math and shit so if that's not what you care about then please scroll past.**
<details>
  <summary>Warning: Arcane is about to talk about the math</summary>
Alright.
So this adds a new stock_market subsystem, which fires once every 20 seconds. I'm still fairly new to subsystem design so I'm probably going to need some feedback on cleaning this up to make it look nicer and run smoother.

So we have 4 associated lists, each attached to the relevant traded datum; this tracks prices, market trends, how long that trend is going to last in SS fires, and market quantity.

Prices fluctuate between 0.5x and 3x the material's single sheet value. This could be tweaked even farther in the future, but for now I'm keeping it at this nice clean margin. Prices fluctuate based on a gaussian normal distribution that is centered on different points based on their **trend**. Upward trending materials are centered in such a way that they'll almost certainly go upwards, but being that it's based on a random chance, not always. Vis versa, downward trends should tend to lose value at about the same rate. We also change our rounding based on this trend data, in order to prevent low price values like iron and glass from getting stuck in the same value or freefall drop over time just because of rounding down. Similarly, neutral trending materials will not change nearly as much, and will generally stay at around the same amount.

When buying or selling a material, the quantity of that material will change on the market. The magnitude of that change depends on how much of that material currently exists on the market. Buying a low quantity material like diamond for example will tangibly increase that material's cost, while buying stacks and stacks of iron and glass won't do much damage to the price of iron as there's usually around 500 full stacks on the market to start with. It's applied at a relative percentage of (qty_changed / new_total_on_market) * price of goods bought/sold at.

In addition to that, there are random "market events" that can occur randomly, which you might miss if you don't follow the newscaster economy news. These three events are fairly barebones now, but in general they have a 1% chance of happening per material, and can more dramatically increase, decrease, or fully reset the value of a material on the market back to it's standard value. This opens up for doing more with it in the future, but that's a later problem.
</details>

## Why It's Good For The Game

This independently resolves some issues related to #78040, that being that lowpop stations, or shifts with few miners would have a new way to be able to still get some access to minerals in a given round. This also provides a unique minigame and alternative way to acquire money in a given shift, using minerals.

**"But Arcane,** I hear you ask. **"Isn't this just the same thing you tried doing way back when and then reverted in #50537?**
Well, fuck man, how you doing I haven't see you around in forever also no you're completely wrong and here's why

**This PR is contingent on #78040 in it's current iteration.** If we are not proceeding with that PR, I need to heavily rework the numbers on this PR because otherwise this PR produces bicycle levels of profit for very little effort. It's stuff like THIS which is why I'm trying to rework minerals. However, as a stopgap measure, this also provides a massive benefit to round progression and gives the QM and the cargo members the ability to prevent round progress from stalling by buying round-critical resources.

Not to mention, as the GMM is not cargo required, more cargo integrated, it also functions as economy content for the rest of the crew.

Now, in general, there may need to be some numbers tweaks that go along with this pull request to ensure that we still haven't fully fucked the crew and the station alike. But that's what this whole process is for, yeah? And it'll go way, WAY better than the last time.

<details>
<summary>Yeah.</summary>
Probably.

image

</details>

## Changelog

:cl:
add: A new export has arrived in the imports section, the Galactic Materials Market! You can use this to buy and sell minerals for profit or cost, as well as stock you station when you don't have any miners.
add: Minerals can be bought on the market either using the station's cargo budget by cargo crew, or privately by everyone else.
del: Any material stacks that could be bought and sold on the market have been removed from the shop.

/:cl:

### CODE REVIEW

**Comment by moocowswag** [comment]
Why are bluespace crystals on the market if plasma isnt, arn`t bluespace crystals made naturally via the formation of plasma and diamonds and thats why we are able to convert plasma and diamonds into artificial crystals?

**Comment by Timonkeyn** [comment]
> Why are bluespace crystals on the market if plasma isnt, arn`t bluespace crystals made naturally via the formation of plasma and diamonds combined and thats why we are able to convert plasma and diamonds into artificial crystals?

Plasma is incendiary. You should consider yourself lucky that the galactic market is selling blue space that could teleport away at any time. There's no way they will sell innate war crimes.

**Comment by Timonkeyn** [comment]
Speaking of war crimes, what happens if I buy 5 stacks of uranium on unsuspecting cargonians?

**Comment by Elli-Skala** [comment]
What if bananaium was a emag unlock tho

**Comment by nianjiilical** [comment]
emagging the market machine to unlock illegal access to iancoin and blowing the entire budget on a shitty jpeg of a corgi when

**Comment by moocowswag** [comment]
> Plasma is incendiary. You should consider yourself lucky that the galactic market is selling blue space that could teleport away at any time. There's no way they will sell innate war crimes.

My point is if nanotrasen has control of plasma it should easily have control of bs crystals and try to lock down trade of either

also the lethality of a single bs crystal is arguably greater than a single sheet of plasma if you wanna talk about danger given that most assets that nanotrasen cares about are in space where crew can get instantly vacuumed

**Comment by deathrobotpunch** [comment]
this is a draft and the maintainers are already trying to remove it

**Comment by CRITAWAKETS** [comment]
holy merry mother of CEO mindset

in all seriousness though, how much money can be done via stocks trading? If the maximum _volume_ of profit (eg even if you get 1% of extra profit, that can still end up in insane amounts of exponential growth if there's no limit) is too high, then it's quite easy for cargo to stop doing it's job and just start investing in the splee- URANIUM market.

**Comment by Bluedino1025** [comment]
> > Plasma is incendiary. You should consider yourself lucky that the galactic market is selling blue space that could teleport away at any time. There's no way they will sell innate war crimes.
> 
> My point is if nanotrasen has control of plasma it should easily have control of bs crystals and try to lock down trade of either , some of the most advanced experimental tech on the station uses bs crystals for a reason
> 
> also the lethality of a single bs crystal is arguably greater than a single sheet of plasma if you wanna talk about danger given that most assets that nanotrasen cares about are in space where crew can get instantly vacuumed

only artificial bs crystals are plasma and diamond, actual bs crystals can be mined and terragov has the most of them so it makes sense.

**Comment by moocowswag** [comment]
> only artificial bs crystals are plasma and diamond, actual bs crystals can be mined and terragov has the most of them so it makes sense.

i thought the minable ones only form in places that  have plasma and diamonds

**Review by Wallemations — APPROVED** [review]
I know it's just a recolor but I think representing your current stock portfolio as a glowing pink holographic cube is really funny

**Comment by TheVekter** [comment]
> This PR is contingent on https://github.com/tgstation/tgstation/pull/78040 in it's current iteration.

Not to be "that guy", but this should probably be a draft until 78040 is merged.

File: tgui/packages/tgui/interfaces/MatMarket.js
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/MatMarket.js:36]: not sure if this works here, but also i'd like to see only stack used unless you had an explicit reason for flex (yes you can nest stacks if that was the reason). the equivalent would be `Stack fill`

File: tgui/packages/tgui/interfaces/MatMarket.js
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/MatMarket.js:4]:
```suggestion

`type Data = {
    orderingPrive: BooleanLike; // you will need to import this
    canOrderCargo: BooleanLike;
    creditBalance: number;
    materials: Material[];
};

type Material = {
    name: string;
    quantity: number;
    id: string; // correct this if its a number
    trend: string;
};
    
```

when you're done change the extension to tsx
.js files should also be .jsx but the bundler accepts the difference

File: tgui/packages/tgui/interfaces/MatMarket.js
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/MatMarket.js:8]:
```suggestion
  const { act, data } = useBackend<Data>(context);   // this will tell your editor that data is the type listed above
  
  const { orderingPrive, canOrderCargo, creditBalance, materials = [] } = data; // better to destructure here (style nit)
```

File: tgui/packages/tgui/interfaces/MatMarket.js
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/MatMarket.js:18]:
```suggestion
              color={!!orderingPrive && !!canOrderCargo ? 'green' : null}
```

File: tgui/packages/tgui/interfaces/MatMarket.js
Comment (jlsnow301) [inline tgui/packages/tgui/interfaces/MatMarket.js:31]:
```suggestion
            <b>{creditBalance || 'zero'}</b> cr.
```

**Review by jlsnow301 — CHANGES_REQUESTED** [review]

**Comment by Comxy** [comment]
Great idea to bring the stock market back, looking forward to gamble all cargos money away.

**Comment by CCC23** [comment]
> holy merry mother of CEO mindset
> 
> in all seriousness though, how much money can be done via stocks trading? If the maximum _volume_ of profit (eg even if you get 1% of extra profit, that can still end up in insane amounts of exponential growth if there's no limit) is too high, then it's quite easy for cargo to stop doing it's job and just start investing in the splee- URANIUM market.

finally time to invest in BLUESPACE BABY WOOOHH BLUESPACE TO THE MOON!!

**Comment by ArcaneMusic** [comment]
image

Tweaks from @jlsnow301 's review as well as some extra extra changes to condense down the UI.

**Comment by necromanceranne** [comment]
putting all my credits into blongocoins

**Comment by CCC23** [comment]
BUY BLUE BUY BLUE SELL SELL SELL

**Comment by ArcaneMusic** [comment]
As a result of needing to rework the core mining concept this was designed for, I'm going to be closing this while I attempt to clean that up, while at the same time incorporating some fixes and feedback to make stocks fun, but not "bike in 90 minutes" busted. Thank you for your patience in advance.

**Pull request closed without merging**
